engine/consensus/approvals/assignment_collector_statemachine_test.go

```diff
@@ -0,0 +1,107 @@
+package approvals
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gammazero/workerpool"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// AssignmentCollectorStateMachineTestSuite is a test suite for testing AssignmentCollectorStateMachine. Contains a minimal set of
+// helper mocks to test the behavior.
+type AssignmentCollectorStateMachineTestSuite struct {
+	BaseAssignmentCollectorTestSuite
+	collector *AssignmentCollectorStateMachine
+}
+
+func TestAssignmentCollectorStateMachine(t *testing.T) {
+	suite.Run(t, new(AssignmentCollectorStateMachineTestSuite))
+}
+
+func (s *AssignmentCollectorStateMachineTestSuite) SetupTest() {
+	s.BaseAssignmentCollectorTestSuite.SetupTest()
+
+	s.collector = NewAssignmentCollectorStateMachine(AssignmentCollectorBase{
+		base: base{
+			workerPool:                           workerpool.New(4),
+			assigner:                             s.assigner,
+			state:                                s.state,
+			headers:                              s.headers,
+			verifier:                             s.sigVerifier,
+			seals:                                s.sealsPL,
+			approvalConduit:                      s.conduit,
+			requestTracker:                       s.requestTracker,
+			requiredApprovalsForSealConstruction: 5,
+		},
+		executedBlock: &s.Block,
+		result:        s.IncorporatedResult.Result,
+		resultID:      s.IncorporatedResult.Result.ID(),
+	})
+}
+
+// TestChangeProcessingStatus_CachingToVerifying tests that state machine correctly performs transition from CachingApprovals to
+// VerifyingApprovals state. After transition all caches approvals and results need to be applied to new state.
+func (s *AssignmentCollectorStateMachineTestSuite) TestChangeProcessingStatus_CachingToVerifying() {
+	require.Equal(s.T(), CachingApprovals, s.collector.ProcessingStatus())
+	results := make([]*flow.IncorporatedResult, 0, 3)
+	approvals := make([]*flow.ResultApproval, 0, s.Chunks.Len())
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		for i := range results {
+			block := unittest.BlockHeaderWithParentFixture(&s.Block)
+			s.blocks[block.ID()] = &block
+			result := unittest.IncorporatedResult.Fixture(
+				unittest.IncorporatedResult.WithIncorporatedBlockID(block.ID()),
+				unittest.IncorporatedResult.WithResult(s.IncorporatedResult.Result),
+			)
+			results[i] = result
```

Comment:
It's better to prepare the results and approvals outside of the go routine, and let the go routine only calls the functions.